mlflow/types/type_hints.py

```diff
@@ -0,0 +1,297 @@
+                    required=True,
+                )
+        if origin_type is dict:
+            if len(args) == 2:
+                if args[0] != str:
+                    raise MlflowException.invalid_parameter_value(
+                        f"Dictionary key type must be str, got {args[0]} in type hint {type_hint}"
+                    )
+                return ColSpecType(
+                    dtype=Map(_infer_colspec_type_from_type_hint(type_hint=args[1]).dtype),
+                    required=True,
+                )
+            raise MlflowException.invalid_parameter_value(
+                f"Dictionary type hint must contain two internal types, got {type_hint}"
+            )
+        if origin_type == Union:
+            if type(None) in args:
+                # This case shouldn't happen, but added for completeness
+                if len(args) < 2:
+                    raise MlflowException.invalid_parameter_value(
+                        f"Union type hint must contain at least one non-None type, got {type_hint}"
+                    )
+                # Optional type
+                elif len(args) == 2:
+                    effective_type = next(arg for arg in args if arg is not type(None))
+                    return ColSpecType(
+                        dtype=_infer_colspec_type_from_type_hint(effective_type).dtype,
+                        required=False,
+                    )
+                # Optional Union type
+                else:
+                    _logger.warning(
+                        "Union type hint with multiple non-None types is inferred as AnyType, "
+                        "and MLflow doesn't validate the data against its internal types."
+                    )
+                    return ColSpecType(dtype=AnyType(), required=False)
+            # Union type with all valid types is matched as AnyType
+            else:
+                _logger.warning(
+                    "Union type hint is inferred as AnyType, and MLflow doesn't validate the data "
+                    "against its internal types."
+                )
+                return ColSpecType(dtype=AnyType(), required=True)
+    else:
+        _invalid_type_hint_error(type_hint)
+
+
+def _invalid_type_hint_error(type_hint: type[Any]) -> None:
+    if type_hint in (list, dict, Optional, Union):
+        raise MlflowException.invalid_parameter_value(
+            f"Unsupported type hint `{type_hint}`, it must include a valid internal type."
+        )
+    raise MlflowException.invalid_parameter_value(
+        f"Unsupported type hint `{type_hint}`, supported types are: "
+        f"{list(TYPE_HINTS_TO_DATATYPE_MAPPING.keys())}, pydantic BaseModel subclasses, "
+        "lists and dictionaries of primitive types, or typing.Any."
+    )
+
+
+def _infer_fields_from_pydantic_model(
+    model: pydantic.BaseModel, field_type: Literal["Property", "ColSpec"]
+) -> list:
```

Comment:
```suggestion
) -> list[???]:
```

to avoid an ambiguous type hint.